packages/shared/src/permissions/permissions.utils.ts

```diff
@@ -78,17 +78,17 @@ export function getReadAccessFilter(userPermissions: UserPermissions) {
 }
 export function hasReadAccess(
   filter: ReadAccessFilter,
-  projects: string | string[] | null | undefined
+  projects: string | string[]
 ): boolean {
-  // If the resource is not project-specific, then everyone has read access
-  if (!projects || !projects.length) {
+  // If the resource is available to all projects (an empty array), then everyone should have read access
+  if (Array.isArray(projects) && !projects.length) {
     return true;
   }
 
   const hasGlobaReadAccess = filter.globalReadAccess;
 
-  // if the user doesn't have project specific roles, fallback to user's global role
-  if (!filter.projects.length) {
+  // if the user doesn't have project specific roles or resource doesn't have a project (project is an empty string), fallback to user's global role
+  if (!filter.projects.length || projects === "") {
```

Comment:
Could we not accomplish the same by doing the following? That way we could keep projects as `string | string[] | null | undefined`, and avoid having to pass in an empty string everywhere.

```suggestion
  if (!filter.projects.length || !projects) {
```

Reply:
We're treating an empty string differently than we're treating an empty array, which is why we've gone the route of not allowing `projects` to be `undefined`.

If we were to introduce the change you're suggesting, we'd lose the ability to do this differentiation.

Some background:
Some resources (metrics, data sources, fact tables, SDK connections) use a projects array as sort of a filter. If empty, there's no filter (all projects). If specified, it's filtering down to only the listed projects.

Other resources (feature flags, experiments) use a singular project field. It can either be undefined or an empty string (not in a project) or set to a specific id.

So being able to differentiate whether it's an empty string or an empty array allows us to control the read access filtering more precisely.

Reply:
Could you double-check? An empty array and `undefined` are not equal in value so I still think what I suggested would work.

```
console.log(Boolean([])); // true
console.log(Boolean(undefined)) // false
```

Reply:
Sorry - I think I misunderstood your request. I initially thought you were asking me to replace both of the following instances

```
hasReadAccess(...., resource.projects || [])
```

AND

```
hasReadAccess(..., resource.project || "")

So that we never had to use the ` || []` or `|| ""` syntax. In that case, we'd not be able to differentiate whether it was undefined because the projects array was undefined or the project string was undefined.

But, correct me if I'm wrong, you were only asking me to update the `hasReadAccess` signature so we didn't have to pass in an empty string?

In that case, I've pushed up this [commit](https://github.com/growthbook/growthbook/pull/2076/commits/978e849899b8eb1d4e84d032bf4cb88cc23dc05e) with those changes. Let me know if that's what you intended.

Reply:
Yep, my comment was just regarding the signature of `hasReadAccess`.